mountpoint-s3/src/cli.rs

```diff
@@ -812,54 +823,90 @@ where
-            Some(max_size_in_mib) => Some(CacheLimit::TotalSize {
+            Some(max_size_in_mib) => CacheLimit::TotalSize {
                 max_size: (max_size_in_mib * 1024 * 1024) as usize,
-            }),
-            None => Some(CacheLimit::default()),
+            },
+            None => CacheLimit::default(),
         };
-        if let Some(cache_limit) = cache_limit {
-            let cache_config = DiskDataCacheConfig {
-                block_size: args.cache_block_size_in_bytes(),
-                limit: cache_limit,
-            };
-            let cache_key = env_unstable_cache_key();
-            let managed_cache_dir = ManagedCacheDir::new_from_parent_with_cache_key(path, cache_key)
-                .context("failed to create cache directory")?;
-
-            let cache = DiskDataCache::new(managed_cache_dir.as_path_buf(), cache_config);
-            let prefetcher = caching_prefetch(cache, runtime, prefetcher_config);
-            let mut fuse_session = create_filesystem(
-                client,
-                prefetcher,
-                &args.bucket_name,
-                &args.prefix.unwrap_or_default(),
-                filesystem_config,
-                fuse_config,
-                &bucket_description,
-            )?;
-
-            fuse_session.run_on_close(Box::new(move || {
-                drop(managed_cache_dir);
-            }));
-
-            return Ok(fuse_session);
-        }
-    }
+        let cache_config = DiskDataCacheConfig {
+            block_size: args.cache_block_size_in_bytes(),
+            limit: cache_limit,
+        };
+        let cache_key = env_unstable_cache_key();
+        let managed_cache_dir = ManagedCacheDir::new_from_parent_with_cache_key(path, cache_key)
+            .context("failed to create cache directory")?;
 
-    if let Some(express_bucket_name) = args.cache_express_bucket_name() {
+        let cache_dir_path = managed_cache_dir.as_path_buf();
+        Ok((managed_cache_dir, DiskDataCache::new(cache_dir_path, cache_config)))
+    };
+
+    let create_express_cache = |args: &CliArgs, client| {
+        let express_bucket_name = args.cache_express_bucket_name().expect("express bucket must be set");
+        // bypass express cache for objects larger than 1 MiB
+        const EXPRESS_CACHE_MAX_OBJECT_SIZE: usize = 1024 * 1024;
         // The cache can be shared across instances mounting the same bucket (including with different prefixes)
         let source_description = &args.bucket_name;
-        let cache = ExpressDataCache::new(
+        ExpressDataCache::new(
             express_bucket_name,
-            client.clone(),
+            client,
             source_description,
             args.cache_block_size_in_bytes(),
-        );
+            EXPRESS_CACHE_MAX_OBJECT_SIZE,
+        )
+    };
+
+    let cache_size_positive = !matches!(args.max_cache_size, Some(0));
+    if args.cache.is_some() && cache_size_positive && args.cache_express_bucket_name().is_some() {
```

Comment:
Does a `match` block work here? It might be more readable

Reply:
Agree with the match. Maybe we should have a method n `CliArgs` that returns an enum with the various cases. Something like:
```
NoCache,
DiskCache(DiskDataCacheConfig),
ExpressCache(ExpressCacheConfig),
MultiLevelCache(DiskDataCacheConfig, ExpressCacheConfig),
```

Reply:
Seems like something like this will work:
```rust
let cache_size_positive = !matches!(args.max_cache_size, Some(0));
match (&args.cache, args.cache_express_bucket_name(), cache_size_positive) {
    (None, Some(express_bucket_name), true) | (None, Some(express_bucket_name), false) | (Some(_), Some(express_bucket_name), false) => todo!(), // express
    (Some(path), None, true) => todo!(), // disk
    (Some(path), Some(express_bucket_name), true) => todo!(), // multilevel
    _ => (), // no cache
}
```

I don't think enum would make it much simpler / understandable, but let me know if you disagree.

Reply:
This is really not great: `let cache_size_positive = !matches!(args.max_cache_size, Some(0));`

An intermediate option could be to have a `CliArgs::disk_data_cache_config(&self) -> Option<DiskDataCacheConfig>` method and then use a match like you suggest.

Reply:
Added a `disk_data_cache_config` and a match block as described in the comment.